lib/i18n/backend/lazy_load.rb

```diff
@@ -0,0 +1,92 @@
+    # implementation avoids loading all translations up front. Instead, it only
+    # loads the translations that belong to the current locale. This offers a
+    # performance incentive for applications which have many translations for
+    # many different locales, but only refer to a single locales' translations
+    # at a time (ex. A Rails workload).  The implementation identifies which
+    # translation files from the load path belong to the current locale by
+    # pattern matching against their path name.
+    #
+    # Specifically, a translation file is considered to belong to a locale if:
+    # a) the filename is in the I18n load path
+    # b) the filename ends in a supported extension (ie. .yml, .json, .po, .rb)
+    # c1) the filename starts with the locale identifier
+    #    OR
+    # c2) the path contains a component equal to the locale identifier
+    #
+    # Examples:
+    # Valid files that will be selected by this backend:
+    #
+    # "files/locales/en-translation.yml" (Selected for locale "en")
+    # "files/locales/fr/translation.po"  (Selected for locale "fr")
+    #
+    # Invalid files that won't be selected by this backend:
+    #
+    # "files/locales/translation-file"
+    # "files/locales/en-translation.unsupported"
+    # "files/locales/french/translation.yml"
+    #
+    # The implementation uses this assumption to defer the loading of
+    # translation files until the current locale actually requires them.
+    #
+    # To use the LazyLoad backend instantiate it and set it to the I18n module.
+    # You can configure lazy loaded backends through the initializer or backends
+    # accessor:
+    #
+    #   # configure I18n backend to use LazyLoad Backend
+    #   I18n.backend = I18n::Backend::LazyLoad.new
+    #
+    class LazyLoad < Simple
+      # Returns whether the current locale is initialized.
+      def initialized?
+        initialized_locales[I18n.locale]
+      end
+
+      # Clean up translations and uninitialize all locales.
+      def reload!
+        @initialized_locales = nil
+        @translations = nil
+      end
+
+      protected
+
+      # Load translations from files that belong to the current locale.
+      def init_translations
+        load_translations(filenames_for_current_locale)
+        initialized_locales[I18n.locale] = true
+      end
+
+      def initialized_locales
+        @initialized_locales ||= Hash.new(false)
+      end
+
+      # Select all files from I18n load path that belong to current locale.
+      # These files must start with the locale identifier (ie. "en", "fr").
+      # or contain a path component equal to the locale identifier (ie. /locales/en/translation.yml)
+      def filenames_for_current_locale
+        I18n.load_path.flatten.select do |path|
+          (basename_starts_with_locale?(path) || path_contains_locale_identifier?(path)) &&
+          supported_extension?(path)
+        end
```

Comment:
I wonder if it would be worthwhile to track which locale files have already been loaded, to prevent them from being loaded twice. Both as a safeguard and as an edge-case perf optimization.

e.g. something like this:

```diff
       def reload!
         @initialized_locales = nil
         @translations = nil
+        @loadable_paths = nil
       end
 
+      def load_translations(*filenames)
+        @loadable_paths = loadable_paths - filenames.flatten
+        super
+      end
+
       protected
 
       # Load translations from files that belong to the current locale.
@@ -64,11 +70,15 @@ module I18n
         @initialized_locales ||= Hash.new(false)
       end
 
+      def loadable_paths
+        @loadable_paths ||= I18n.load_path.flatten
+      end
+
       # Select all files from I18n load path that belong to current locale.
       # These files must start with the locale identifier (ie. "en", "fr").
       # or contain a path component equal to the locale identifier (ie. /locales/en/translation.yml)
       def filenames_for_current_locale
-        I18n.load_path.flatten.select do |path|
+        loadable_paths.select do |path|
           (basename_starts_with_locale?(path) || path_contains_locale_identifier?(path)) &&
           supported_extension?(path)
         end
```

So basically when you load translations, you remove those paths from `loadable_paths` and even if there is some overlap with another locale, you don't load them again unless the backend has been reloaded.

I don't want to add complexity if unnecessary so this is just an idea.

Reply:
Yeah I think having this kind of safeguard would allow us to be more loose with formats, so e.g. my earlier point about `pt.yml` vs `pt-BR.yml` wouldn't be such a big deal if this were in place: `pt` would load both, but if you switched to `pt-BR` it would not load that locale's translations again.

Reply:
I've removed the latter format (ie. locale in path components), and enforced the presence of a demarcation ("_") if text proceeds the locale.

This simplifies the assumptions, ie.
1. We can infer `#available_locales` concretely
2. We don't need to safeguard against multiple loads (although this solution was 👍 )

Thanks for your help.